### PR TITLE
[FSSDK-10619] Refactor project config manager to be injectable

### PR DESCRIPTION
**Pull request by raju-opti**
## Summary
- This PR allows the user to create a ProjectConfigManager and pass it directly to createInstance. It exposes two factory methods `createStaticProjectConfigManager` and `createPollingProjectConfigManager` for this purpose.

## Test plan
- New tests have been added and relevant old tests has been updated

## Issues
- FSSDK-10619
- FSSDK-10639
- FSSDK-10641

### CODE REVIEW

**Comment by mikechu-optimizely** [comment]
Sorry I didn't get to this today.  It's first on my list Mon

**Review by mikechu-optimizely — APPROVED** [review]
LGTM.

File: lib/project_config/project_config.ts
Comment (junaed-optimizely) [inline lib/project_config/project_config.ts:1]: Why are we renaming it to `project_config` from`index` ?

File: lib/optimizely/index.ts [hunk 1]
```diff
@@ -41,15 +41,15 @@ import {
 } from '../shared_types';
 import { newErrorDecision } from '../optimizely_decision';
 import OptimizelyUserContext from '../optimizely_user_context';
-import { createProjectConfigManager, ProjectConfigManager } from '../core/project_config/project_config_manager';
+// import { createProjectConfigManager, ProjectConfigManager } from '../core/project_config/project_config_manager';
```
Comment (junaed-optimizely) [inline lib/optimizely/index.ts:44]: Comment that can be removed!?

**Review by junaed-optimizely — APPROVED** [review]
Great work! So many great additions + modifications. 
LGTM!
I just have couple of curious questions.

**Review by jaeopt — APPROVED** [review]
LGTM with a few comments.
PR is too big and includes completely rewritten codes, so hard to review. I see no clear errors, but expect all these changes covered by unit tests (and later with FSC).

File: lib/optimizely/index.ts [hunk 2]
```diff
@@ -1262,7 +1258,7 @@ export default class Optimizely implements Client {
       if (!configObj) {
         return null;
       }
-      return this.projectConfigManager.getOptimizelyConfig();
+      return this.projectConfigManager.getOptimizelyConfig() || null;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/optimizely/index.ts:1261]: Yeah. Good catch. I'll return undefined from both.

File: lib/project_config/polling_datafile_manager.ts [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  private async syncDatafile(): Promise<void> {
+    const headers: Headers = {};
+    if (this.lastResponseLastModified) {
+      headers['if-modified-since'] = this.lastResponseLastModified;
+    }
+
+    if (this.datafileAccessToken) {
+      this.logger?.debug('Adding Authorization header with Bearer Token');
+      headers['Authorization'] = `Bearer ${this.datafileAccessToken}`;
+    }
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/project_config/polling_datafile_manager.ts:190]: moved out to a separate function

File: lib/project_config/project_config.ts
Comment (raju-opti) [reply lib/project_config/project_config.ts:1]: index have the connotation that it is an entry point for a module. But here, project_config directory is home to multiple modules related to project_config. we don't export all of them from index. For example: modules for project config manager and such. That's why the renaming

File: lib/optimizely/index.ts [hunk 4]
```diff
@@ -192,7 +188,7 @@ export default class Optimizely implements Client {
    * Returns the project configuration retrieved from projectConfigManager
    * @return {projectConfig.ProjectConfig}
    */
-  getProjectConfig(): projectConfig.ProjectConfig | null {
+  getProjectConfig(): projectConfig.ProjectConfig | undefined {
```
[the remark this answers is not in the document]
Comment (raju2592) [reply lib/optimizely/index.ts:191]: Let's update both to return undefined?

File: lib/project_config/config_manager_factory.react_native.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { getPollingConfigManager, PollingConfigManagerConfig } from "./config_manager_factory";
+import { BrowserRequestHandler } from "../utils/http_request_handler/browser_request_handler";
+import { ProjectConfigManager } from "./project_config_manager";
+import ReactNativeAsyncStorageCache from "../plugins/key_value_cache/reactNativeAsyncStorageCache";
+
+export const createPollingProjectConfigManager = (config: PollingConfigManagerConfig): ProjectConfigManager => {
+  const defaultConfig = {
+    autoUpdate: true,
+    requestHandler: new BrowserRequestHandler(),
+    cache: new ReactNativeAsyncStorageCache(),
```
[the remark this answers is not in the document]
Comment (raju2592) [reply lib/project_config/config_manager_factory.react_native.ts:26]: In the previous implementation, only node version used the authenticated url template. The datafile manager class uses the default template if not provided. I see as we moved the authentication header in the common datafile manager class, url template selection logic can move there. I'll refactor accordingly.

File: lib/optimizely/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/optimizely/index.ts:191]: Both methods returns undefiend in the projectConfigManager. I am looking to use undefined for missing values. This also works well with typescript optional properties. But we should have consistent naming.

I see the many of the methods of the optimizly class already returns null, I'll use null for this as well.

File: lib/project_config/config_manager_factory.react_native.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/project_config/config_manager_factory.react_native.ts:26]: In the previous implementation, only node version used the authenticated url template. The datafile manager class uses the default template if not provided. I see as we moved the authentication header in the common datafile manager class, url template selection logic can move there. I'll refactor accordingly.

[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/project_config/config_manager_factory.react_native.ts:26]: moved this logic to the PollingDatafileManager class

**Pull request merged**